typing/parmatch.ml

```diff
@@ -2363,18 +2355,31 @@ let rec matrix_stable_vars rs = match rs with
     Vars (List.fold_left IdSet.union IdSet.empty stables_in_varsets)
 | rs ->
     let rs = simplify_first_col rs in
-    if not (all_coherent (first_column rs)) then All
+    if not (all_coherent (first_column rs))
+    (* If the column is ill-typed but deemed coherent, we might spuriously
+       warn about some variables being unstable.
+       As sad as that might be, the warning can be silenced by splitting the
+       or-pattern...
+    *)
+    then All
     else begin
-      (* If the column is ill-typed but deemed coherent, we might spuriously
-         warn about some variables being unstable.
-
-         As sad as that might be, the warning can be silenced by splitting the
-         or-pattern...
-      *)
-      let submatrices = split_rows rs in
-      let submat_stable = List.map matrix_stable_vars submatrices in
-      (* a stable variable must be stable in each submatrix *)
-      List.fold_left stable_inter All submat_stable
+      let extend_row columns r =
+        { r with row = columns @ r.row } in
+      let q0 = discr_pat omega rs in
+      match build_specialized_submatrices ~extend_row q0 rs with
+      | { default; constrs = [] } ->
+          (* the first column contains no head constructor;
+             they are all _ after simplification, so it can be dropped *)
+          matrix_stable_vars default
+      | { default = _; constrs } ->
+          (* A stable variable must be stable in each submatrix.
+             If the first column contains some head constructors, there
+             is no need to look at stability for the default matrix: all
+             other submatrices contain the default matrix, so they have
+             less stable variables. *)
```

Comment:
There are two unrelated parts to this comment, personally I'd like it better if there was a visual hint that that's the case (i.e. I'd add a blank line). When there is only one paragraph I have to reach the end to realize that "oh, ok he's talking about something unrelated".

That being said, I don't think the second part of that comment is that useful: the operation (building specialized matrices) is frequent in parmatch, and the doc of `build_specialized_submatrices` already says what you're repeating here.

Reply:
Hum, OK:
- github doesn't display the whole comment on the main interface; when I talk about "the second part" I mean "everything github is showing right now".
- I had actually missed the last bit of that comment (because of the "oh ok he's just repeating things we already know" mentioned previously). 
Maybe you could shorten the comment to:
  > A stable variable must be stable in each submatrix.
  > N.B. all the [constrs] submatrices contain the default matrix, so they have less stable variables.

Reply:
I added a line between the two parts. I would like to keep the long fully-explicit version for now, because the reasoning may change with negative rows and it's convenient to have the full story to think. I think I will have to edit those comments again anyway.